modules/renter/download.go

```diff
@@ -700,7 +707,18 @@ func (dw *DownloadFileWriter) Destination() string {
 
 // WriteAt writes the passed bytes at the specified offset.
 func (dw *DownloadFileWriter) WriteAt(b []byte, off int64) (int, error) {
-	return dw.f.WriteAt(b, off-int64(dw.offset))
+	if dw.written+uint64(len(b)) > dw.length {
+		build.Critical("DownloadFileWriter write exceeds file length")
+	}
+	n, err := dw.f.WriteAt(b, off-int64(dw.offset))
+	if err != nil {
+		return n, err
+	}
+	dw.written += uint64(n)
+	if dw.written >= dw.length {
+		return n, dw.f.Close()
+	}
+	return n, err
```

Comment:
I think that calling `Close` inside of the `WriteAt` function is the wrong move. Whatever is calling `WriteAt` should be the method calling `Close`, which would more appropriately match the other interfaces in the golang language.